MalomAPI_cpp/perfect_wrappers.h

```diff
@@ -0,0 +1,362 @@
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef PERFECT_WRAPPER_H_INCLUDED
+#define PERFECT_WRAPPER_H_INCLUDED
+
+#include "perfect_common.h"
+#include "perfect_hash.h"
+#include "perfect_symmetries.h"
+#include "perfect_debug.h"
+#include "perfect_sector.h"
+#include "perfect_sector_graph.h"
+
+#include <cassert>
+#include <cmath> // for factorial function
+#include <iostream>
+#include <map>
+#include <set>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+namespace Wrappers {
+
+class WSector;
+
+extern std::unordered_map<id, int> sector_sizes;
+static int f_inv_count[] {1,     4,     30,    158,    757,    2830,  8774,
+                          22188, 46879, 82880, 124124, 157668, 170854};
+
+struct WID
+{
+    int W, B, WF, BF;
+    WID(int W, int B, int WF, int BF)
+        : W(W)
+        , B(B)
+        , WF(WF)
+        , BF(BF)
+    { }
+    WID(id id)
+        : W(id.W)
+        , B(id.B)
+        , WF(id.WF)
+        , BF(id.BF)
+    { }
+    ::id tonat() { return ::id(W, B, WF, BF); }
+    void negate();
+    WID operator-(WID s);
+
+    std::string ToString() { return this->tonat().to_string(); }
+
+    int GetHashCode() { return (W << 0) | (B << 4) | (WF << 8) | (BF << 12); }
+
+private:
+    static int64_t factorial(int n)
+    {
+        if (n == 0)
+            return 1;
+        else
+            return n * factorial(n - 1);
+    }
+
+    static int64_t nCr(int n, int r)
+    {
+        return factorial(n) / (factorial(r) * factorial(n - r));
+    }
+
+public:
+    int size()
+    {
+        auto tn = tonat();
+        if (sector_sizes.count(tn) == 0) {
+            sector_sizes[tn] = static_cast<int>(nCr(24 - W, B)) *
+                               f_inv_count[W];
+        }
+        return sector_sizes[tn];
+    }
+
+    bool operator==(const WID &other) const
+    {
+        return W == other.W && B == other.B && WF == other.WF && BF == other.BF;
+    }
+
+    bool operator<(const WID &other) const
+    {
+        return std::tie(W, B, WF, BF) <
+               std::tie(other.W, other.B, other.WF, other.BF);
+    }
+};
+
+struct eval_elem
+{
+    enum class cas { val, count, sym };
+    cas c;
+    int x;
+
+    eval_elem(cas c, int x)
+        : c(c)
+        , x(x)
+    { }
+    eval_elem(::eval_elem e)
+        : c(static_cast<cas>(e.c))
+        , x(e.x)
+    { }
+};
+
+struct gui_eval_elem2;
+
+class WSector
+{
+public:
+    ::Sector *s;
+    WSector(WID id)
+        : s(new ::Sector(id.tonat()))
+    { }
+
+    std::pair<int, Wrappers::gui_eval_elem2> hash(board a);
+
+    sec_val sval() { return s->sval; }
+};
+
+struct gui_eval_elem2
+{
+private:
+    // azert cannot be valid, because it cannot contain a count (as asserted by
```

Comment:
"azert cannot be valid, because it cannot contain a count"
-->
"could not be simply val instead of sec_val, because val cannot contain a count"